nodeup/pkg/model/hooks.go

```diff
@@ -17,53 +17,130 @@ limitations under the License.
 package model
 
 import (
-	"github.com/golang/glog"
+	"errors"
+	"fmt"
+	"strings"
+
 	"k8s.io/kops/pkg/apis/kops"
 	"k8s.io/kops/pkg/systemd"
 	"k8s.io/kops/upup/pkg/fi"
 	"k8s.io/kops/upup/pkg/fi/nodeup/nodetasks"
-	"strconv"
-	"strings"
+
+	"github.com/golang/glog"
 )
 
-// HooksBuilder configures the hooks
+// HookBuilder configures the hooks
 type HookBuilder struct {
 	*NodeupModelContext
 }
 
 var _ fi.ModelBuilder = &HookBuilder{}
 
-func (b *HookBuilder) Build(c *fi.ModelBuilderContext) error {
-	for i := range b.Cluster.Spec.Hooks {
-		hook := &b.Cluster.Spec.Hooks[i]
-
-		// TODO: Allow (alphanumeric?) names
-		name := strconv.Itoa(i + 1)
-
-		service, err := b.buildSystemdService(name, hook)
-		if err != nil {
-			return err
+// Build is responsible for implementing the cluster hook
+func (h *HookBuilder) Build(c *fi.ModelBuilderContext) error {
+	// we keep a list of hooks name so we can allow local instanceGroup hooks override the cluster ones
+	hookNames := make(map[string]bool, 0)
+	for i, spec := range []*[]kops.HookSpec{&h.InstanceGroup.Spec.Hooks, &h.Cluster.Spec.Hooks} {
+		for _, hook := range *spec {
+			isInstanceGroup := i == 0
+			// filter out on master and node flags if required
+			if (hook.MasterOnly && !h.IsMaster) || (hook.NodeOnly && h.IsMaster) {
+				continue
+			}
+			// i dont want to effect those whom are already using the hooks, so i'm gonna try an keep the name for now
+			// i.e. use the default naming convention - kops-hook-<index>, only those using the Name or hooks in IG should alter
+			var name string
+			switch hook.Name {
+			case "":
+				name = fmt.Sprintf("kops-hook-%d", i)
+				if isInstanceGroup {
+					name = fmt.Sprintf("%s-ig", name)
+				}
+			default:
+				name = hook.Name
+			}
+
+			if _, found := hookNames[name]; found {
+				glog.V(2).Infof("Skipping the hook: %v as we've already processed a similar service name", name)
```

Comment:
The interaction with MasterOnly / NodeOnly is tricky here.

One option would be through validation to exclude Hooks at the InstanceGroup level from setting MasterOnly / NodeOnly.

Do you have a use-case for overriding by Name?  It seems likely to cause confusion...